testsuite/tests/typing-misc/let_rec_disallowed.ml

```diff
@@ -0,0 +1,81 @@
+let rec x = let y = () in x;;
+
+let rec x = let module M = struct let f = x let g = x () end in fun () -> ();;
+
+let rec x = let module M = struct let f = x () let g = x end in fun () -> ();;
+
+let rec x = (let module M = struct let f = y 0 let g = () end in fun () -> ())
+    and y = succ;;
+
+let rec x = let module M = struct module N = struct let y = x end end in M.N.y;;
+
+let rec x = let module M = struct let f = x () and g = x end in fun () -> ();;
+
+class c _ = object end;;
+let rec x = new c x;;
+
+let rec x = ignore x;;
+
+let rec x = y 0 and y _ = ();;
+
+let rec x = [|y|]
+and y = 0;;
+
+let rec c = { c with Complex.re = 1.0 };;
+
+let rec x = { x with contents = 3 };;
+
+let rec b = if b then true else false;;
+
+let r = ref ()
+let rec x = r := x;;
+
+let rec x =
+  for i = 0 to 1 do
+    let z = y in ignore z
+  done
+and y = x; ();;
+
+let rec x =
+  for i = 0 to y do
+    ()
+  done
+and y = 10;;
+
+let rec x =
+  for i = y to 10 do
+    ()
+  done
+and y = 0;;
+
+let rec x =
+  while false do
+    let y = x in ignore y
+  done
+and y = x; ();;
+
+let rec x =
+  while y do
+    ()
+  done
+and y = false;;
+
+let rec x =
+  while y do
+    let y = x in ignore y
+  done
+and y = false
+
+let rec x = y#m and y = object method m = () end;;
+
+let rec x = (object method m _ = () end)#m x;;
+
+let rec x = y.contents and y = { contents = 3 };;
+
+let rec x = object val mutable v = 0 method m = v <- y end and y = 1;;
+
+let rec x = assert y and y = true;;
+
+let rec x = object method m = x end;;
```

Comment:
It is not clear to me why this case is rejected, given that method calls delay evaluation (`object method m = fun () -> x end` is delayed as well). Isn't this just a matter of how object declarations are elaborated in practice, that could change with a refined implementation? If so, then maybe it would make sense to add a comment to say it explicitly.